CRM/MembershipExtras/Form/RecurringContribution/AddNoInstalmentsMembershipLineItem.php

```diff
@@ -0,0 +1,296 @@
+    ]);
+
+    return array_shift($result['values']);
+  }
+
+  private function calculateEndDateForMembership() {
+    if ($this->membershipType->duration_unit == 'lifetime') {
+      return NULL;
+    }
+
+    return $this->lineItemParams['end_date'];
+  }
+
+  private function getDefaultPriceFieldValueForMembershipType($membershipTypeID) {
+    $result = civicrm_api3('PriceFieldValue', 'get', [
+      'sequential' => 1,
+      'membership_type_id' => $membershipTypeID,
+      'price_field_id.price_set_id.name' => 'default_membership_type_amount',
+    ]);
+
+    if ($result['count'] > 0) {
+      return array_shift($result['values']);
+    }
+
+    return [];
+  }
+
+  /**
+   * Gets the parameters that are required to
+   * create one off payment line item.
+   *
+   * @return array
+   */
+  private function getOneOffPaymentLineItemParams() {
+    if ($this->submittedValues['payment_type'] != self::PAYMENT_TYPE_ONE_OFF_PAYMENT) {
+      return [];
+    }
+
+    $amountExcTax = $this->submittedValues['amount_exc_tax'] ?? 0;
+    $amountIncTax = $this->submittedValues['amount_inc_tax'] ?? 0;
+    $taxAmount = $amountIncTax - $amountExcTax;
+
+    return [
+      'unit_price' => $amountExcTax,
+      'line_total' => $amountExcTax,
+      'tax_amount' => $taxAmount,
+      'financial_type_id' => $this->submittedValues['noinstalmentline_financial_type_id'],
+    ];
+  }
+
+  /**
+   * Creates the one-off payment (contribution) if such option
+   * is selected. It also links it to the created membership.
+   *
+   * @param array $createdLineItemData
+   * @return array
+   */
+  private function createOneOffPayment($createdLineItemData) {
+    $totalIncTax = $this->submittedValues['amount_inc_tax'] ?? 0;
+
+    $contribution = civicrm_api3('Contribution', 'create', [
+      'financial_type_id' => $createdLineItemData['line_item']['financial_type_id'],
+      'receive_date' => $this->submittedValues['scheduled_charge_date'],
+      'total_amount' => $totalIncTax,
+      'fee_amount' => 0,
+      'net_amount' => $totalIncTax,
+      'tax_amount' => $createdLineItemData['line_item']['tax_amount'],
+      'contact_id' => $this->recurringContribution['contact_id'],
+      'contribution_status_id' => 'Pending',
+      'currency' => $this->recurringContribution['currency'],
+      'payment_instrument_id' => $this->recurringContribution['payment_instrument_id'],
+      'source' => 'Manage Instalments form',
+      'contribution_recur_id' => $this->recurringContribution['id'],
+      'is_pay_later' => TRUE,
+    ]);
+
+    civicrm_api3('MembershipPayment', 'create', [
+      'membership_id' => $createdLineItemData['membership']['id'],
+      'contribution_id' => $contribution['id'],
+    ]);
+
+    return array_shift($contribution['values']);
+  }
+
+  /**
+   * Sends the confirmation email if such option is selected.
+   *
+   * @param int $contributionId
+   * @return void
+   */
+  private function sendConfirmationEmail($contributionId) {
```

Comment:
Just curious about this, do we really send a confirmation if the contribution is created and set to pending?

Reply:
the specs mentions given the user the option to send an email, I don't have personal preferences on this and it was easy to implement anyway.